### PR TITLE
replace CtfeCompiler with foreachExpAndVar()

### PR DESCRIPTION
**Pull request by WalterBright**
First step towards splitting traversing the AST and incrementing the number of VarDeclarations into separate operations.

### CODE REVIEW

File: src/dmd/dinterpret.d [hunk 1]
```diff
@@ -454,270 +454,227 @@ struct CompiledCtfeFunction
...
+            foreach (ca; *s.catches)
+            {
+                if (ca.var)
+                    dgVar(ca.var);
+                if (ca.handler)
+                    visit(ca.handler);
+            }
+        }
+
+        void visitTryFinally(TryFinallyStatement s)
```
Comment (JinShil) [inline src/dmd/dinterpret.d]: Aren't the names of these functions redundant?  Can't they all be named `visit` and have D's overload rules handle the resolution?

File: src/dmd/dinterpret.d [hunk 2]
```diff
@@ -454,270 +454,227 @@ struct CompiledCtfeFunction
...
-        }
-    }
-
-    override void visit(ReturnStatement s)
-    {
-        debug (LOGCOMPILE)
-        {
-            printf("%s ReturnStatement::ctfeCompile\n", s.loc.toChars());
+            dgExp(s.condition);
+            // Note that the body contains the the Case and Default
```
Comment (atilaneves) [inline src/dmd/dinterpret.d]: s/the the/the/

File: src/dmd/dinterpret.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/dinterpret.d]: Overloading is not supported for nested functions.

**Review by atilaneves — APPROVED** [review]

**Comment by UplinkCoder** [comment]
The main point of this code still being there and the way it works.
Is to prevent broken trees to come out of CTFE.
The `assert(0)` statements in there should be retained.

**Pull request merged**